pysat/_meta.py

```diff
@@ -386,26 +386,66 @@ def __setitem__(self, data_vars, input_dat):
-                                # Empty list, ensure there is something to set
-                                to_be_set = ['']
-                            if isinstance(to_be_set[0], str) \
-                                    or isinstance(to_be_set, bytes):
-                                if isinstance(to_be_set, bytes):
+                        good_set = True
+
+                        # See if this meta data key has already been defined
+                        # in MetaLabels
+                        if ikey in self.labels.label_attrs.keys():
+                            iattr = self.labels.label_attrs[ikey]
+                            if not isinstance(
+                                    to_be_set, self.labels.label_type[iattr]):
+                                # If this is a disagreement between byte data
+                                # and an expected str, resolve it here
+                                if(isinstance(to_be_set, bytes)
+                                   and self.labels.label_type[iattr] == str):
                                     to_be_set = to_be_set.decode("utf-8")
-
-                                self._data.loc[var, ikey] = '\n\n'.join(
-                                    to_be_set)
-                            else:
-                                warnings.warn(' '.join(('Array elements are',
-                                                        'not allowed in meta.',
-                                                        'Dropping input for',
-                                                        var, 'with key', ikey)))
+                                else:
+                                    # This type is incorrect, try casting it
+                                    wmsg = ''.join(['Metadata does not have ',
+                                                    'expected type ',
+                                                    repr(self.labels.label_type[
+                                                        iattr])])
+                                    try:
+                                        if hasattr(to_be_set, '__iter__'):
+                                            if self.labels.label_type[
+                                                    iattr] == str:
+                                                to_be_set = '\n\n'.join(
+                                                    [str(tval) for tval in
+                                                     to_be_set])
+                                            else:
+                                                raise TypeError("can't recast")
+                                        else:
+                                            to_be_set = self.labels.label_type[
+                                                iattr](to_be_set)
+
+                                        # Inform user data was recast
+                                        pysat.logger.warning(''.join((
```

Comment:
I tried this branch out on an ICON file. It leads to a whole long list of warnings.

```
In [8]: ivm = pysat.Instrument('icon', 'ivm', inst_id='a', update_files=True)

In [10]: ivm.load(2020, 1)
pysat WARNING: Metadata does not have expected type <class 'float'>. Recasting input for 'Epoch' with key 'FillVal'
pysat WARNING: Metadata does not have expected type <class 'float'>. Recasting input for 'ICON_L27_Altitude' with key 'FillVal'
pysat WARNING: Metadata does not have expected type <class 'numpy.float64'>. Recasting input for 'ICON_L27_Fractional_Ion_Density_H' with key 'Valid_Min'
pysat WARNING: Metadata does not have expected type <class 'float'>. Recasting input for 'ICON_L27_Fractional_Ion_Density_H' with key 'ValidMax'
pysat WARNING: Metadata does not have expected type <class 'float'>. Recasting input for 'ICON_L27_Fractional_Ion_Density_H' with key 'ValidMin'
pysat WARNING: Metadata does not have expected type <class 'numpy.float64'>. Recasting input for 'ICON_L27_Fractional_Ion_Density_H' with key 'Valid_Max'
```

Only showing a small fraction since it is way too much. Two contrasting requests:
 - From the error I don't know what type the data loaded as. I think adding that info may be helpful.
 - There are too many messages :(

Reply:
Also tried it out on a MIGHTI file as a check to make sure it isn't just my files that produce warnings. It started to produce warnings but ran into an issue.

```
In [16]: mighti = pysat.Instrument('icon', 'mighti', inst_id='vector', tag='vector_wind_green')

In [19]: mighti.load(2020, 1)
pysat WARNING: Metadata does not have expected type <class 'float'>. Recasting input for 'Epoch' with key 'FillVal'
pysat WARNING: Metadata does not have expected type <class 'float'>. Recasting input for 'Epoch' with key 'ValidMin'
pysat WARNING: Metadata does not have expected type <class 'float'>. Recasting input for 'Epoch' with key 'ValidMax'
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-19-9f1312956ee0> in <module>
----> 1 mighti.load(2020, 1)

~/Code/pysat/pysat/_meta.py in __setitem__(self, data_vars, input_dat)
    442                             # Update the MetaLabels object and the existing
    443                             # metadata to ensure all data have all labels
--> 444                             self.labels.update(iattr, ikey, itype)
    445                             self._label_setter(ikey, ikey, type(to_be_set))
    446 

~/Code/pysat/pysat/_meta.py in update(self, lattr, lname, ltype)
   1821         else:
   1822             # This is an invalid meta data type, raise a TypeError
-> 1823             raise TypeError(''.join(['iterable types like ', repr(ltype),
   1824                                      ' (set for ', repr(lattr),
   1825                                      ') are not allowed']))

TypeError: iterable types like <class 'numpy.ndarray'> (set for 'Valid_Range') are not allowed

```

Reply:
So, this is foreseen.  Now that there's a type check, the arrays that can sneak into meta when you first initialize the data are caught.  This should be rectified separately in [pysat/pysatNASA#99](https://github.com/pysat/pysatNASA/issues/99).

Reply:
> There are too many messages :(

I could put the messages at the info level.  Since the data still exists and is loaded, that could be seen as appropriate.

Reply:
> So, this is foreseen.

Right... I missed your note about that at the top. eek! My bad :(

> I could put the messages at the info level. Since the data still exists and is loaded, that could be seen as appropriate.

Info is definitely better, thinking maybe even debug.

Reply:
I don't want to put it at DEBUG because it would be better if people used the correct type of data, thus avoiding potential casting errors.  But I am fine with INFO, because it's just not that dire. 🐺

Reply:
I fully agree that it is best to use the correct types of data. However, the file creation process is separated from pysat and we are unlikely to get any existing datasets changed. I thought I had the correct types for IVM but I'm getting pages of warnings every load. The log will be full of items people can't do anything about. 

I'm curious what the issue is for the IVM files. For some metadata I did pass through what I was given from earlier files but most of it was set within pysat. v2.x pysat, but still pysat.